fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -823,6 +828,92 @@ public void buildIndex() {
         asyncToSync(buildIndexAsync());
     }
 
+
+    /**
+     * Build an index across multiple transactions in parallel (if needed).
+     * <p>
+     * This method assumes it need to mark the index as readable after building the index. But it will not mark if the
+     * index is built in parallel. It is the caller's responsibility to check all subtasks are completed and then mark
+     * the index as readable.
+     * <p>
+     * It is running in an synchronous context because the cursor returned from {@link FDBRecordStore#getPrimaryKeyBoundaries(Tuple, Tuple)} is blocking.
+     * <p>
+     * The caller is supposed to define how they want to build the range. For example, an upper layer may want to queue
+     * a job for each range to be built so the jobs can be executed and re-tried as needed.
+     *
+     * @param minParallelism the minimum number of substasks to enable Parallelism
+     * @param maxParallelism the maximum number of substasks generated
+     * @param buildRangeProcessor build index for entries between given the primary key endpoints (first inclusive,
+     *     second exclusive) and mark the index as readable when all subtasks are done
+     */
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    public void buildIndexInMultipleJobs(int minParallelism, int maxParallelism,
+                                         Consumer<Pair<Tuple, Tuple>> buildRangeProcessor) {
+        TupleRange tupleRange = runner.asyncToSync(FDBStoreTimer.Waits.WAIT_BUILD_ENDPOINTS, buildEndpoints());
+
+        if (tupleRange != null) {
+            List<Tuple> boundaries = getPrimaryKeyBoundaries(tupleRange);
+            if (boundaries.size() - 1 < minParallelism) {
+                asyncToSync(
+                        buildRange(Key.Evaluated.fromTuple(tupleRange.getLow()), Key.Evaluated.fromTuple(tupleRange.getHigh())));
+            } else {
+                List<Pair<Tuple, Tuple>> tasks = new ArrayList<>();
+                int stepSize = (boundaries.size() + maxParallelism - 1) / maxParallelism; // step size >= 1
+                int start = 0;
+                while (true) {
+                    int next = start + stepSize;
+                    if (next < boundaries.size() - 1) {
+                        tasks.add(Pair.of(boundaries.get(start), boundaries.get(next)));
+                    } else {
+                        tasks.add(Pair.of(boundaries.get(start), boundaries.get(boundaries.size() - 1)));
+                        break;
+                    }
+                }
+                // Shuffle the tasks to avoid hot spots.
+                Collections.shuffle(tasks);
```

Comment:
Presumably if they are being built in parallel, then shuffling really doesn't buy you anything, right?  I guess maybe we cannot make assumptions on how the consumer is handling its input, so maybe this does make some sense though...

Reply:
Removed because of the contract change above.